compiler/rustc_middle/src/ty/inhabitedness/mod.rs

```diff
@@ -1,14 +1,132 @@
+            Self::Or([a, b]) => try_or(a, b, |x| x.apply_inner(tcx, param_env, mod_pred)),
+        }
+    }
+
+    pub fn and(self, tcx: TyCtxt<'tcx>, other: Self) -> Self {
+        match (self, other) {
+            (Self::False, _) | (_, Self::False) => Self::False,
+            (Self::True, a) | (a, Self::True) => a,
+            (Self::Module(a), Self::Module(b)) if a == b => Self::Module(a),
+            (Self::Module(a), Self::Module(b)) if tcx.is_descendant_of(a, b) => Self::Module(a),
+            (Self::Module(a), Self::Module(b)) if tcx.is_descendant_of(b, a) => Self::Module(b),
+            (Self::ConstNonZero(a), Self::ConstNonZero(b)) if a == b => Self::ConstNonZero(a),
+            (a, b) => Self::And(tcx.arena.alloc([a, b])),
+        }
+    }
+
+    pub fn or(self, tcx: TyCtxt<'tcx>, other: Self) -> Self {
+        match (self, other) {
+            (Self::True, _) | (_, Self::True) => Self::True,
+            (Self::False, a) | (a, Self::False) => a,
+            (Self::Module(a), Self::Module(b)) if a == b => Self::Module(a),
+            (Self::Module(a), Self::Module(b)) if tcx.is_descendant_of(a, b) => Self::Module(b),
+            (Self::Module(a), Self::Module(b)) if tcx.is_descendant_of(b, a) => Self::Module(a),
+            (Self::ConstNonZero(a), Self::ConstNonZero(b)) if a == b => Self::ConstNonZero(a),
+            (a, b) => Self::Or(tcx.arena.alloc([a, b])),
+        }
+    }
+
+    pub fn all(tcx: TyCtxt<'tcx>, iter: impl IntoIterator<Item = Self>) -> Self {
+        let mut result = Self::True;
+        for pred in iter {
+            if matches!(pred, Self::False) {
+                return Self::False;
+            }
+            result = result.and(tcx, pred);
+        }
+        result
+    }
+
+    pub fn any(tcx: TyCtxt<'tcx>, iter: impl IntoIterator<Item = Self>) -> Self {
+        let mut result = Self::False;
+        for pred in iter {
+            if matches!(pred, Self::True) {
+                return Self::True;
+            }
+            result = result.or(tcx, pred);
+        }
+        result
+    }
+}
+
+pub(crate) fn uninhabited_predicate<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    ty: Ty<'tcx>,
+) -> UninhabitedPredicate<'tcx> {
+    match *ty.kind() {
+        Adt(adt, substs) => {
+            if let Some(def_id) = adt.did().as_local() {
+                if matches!(tcx.representability(def_id), ty::Representability::Infinite) {
+                    return UninhabitedPredicate::False;
+                }
+            }
+            UninhabitedPredicate::all(
+                tcx,
+                adt.variants()
+                    .iter()
+                    .map(|variant| variant.uninhabited_predicate(tcx, substs, adt)),
+            )
+        }
+
+        Tuple(ref tys) => {
+            UninhabitedPredicate::any(tcx, tys.iter().map(|ty| ty.uninhabited_predicate(tcx)))
+        }
+
+        Array(ty, len) => {
+            // todo is this valid??
```

Comment:
please advise :)

Reply:
That is, can I use `ParamEnv::empty()` to detect simple cases? (like when array length is a literal)

Reply:
I think what you want for that case is `len.kind().ty_to_machine_usize(tcx)`. This will not attempt evaluation, but just grab an already evaluated value

Reply:
Yes perfect.